File: app/src/main/java/com/hjw/fundplan/base/BaseActivity.java
package com.hjw.fundplan.base;


import android.app.Activity;
import android.app.Dialog;
import android.content.Context;
import android.content.DialogInterface;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.View;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.LifecycleOwner;

public abstract class BaseActivity<T extends IContract.IPresenter<? extends IContract.IView>> extends AppCompatActivity implements IContract.IView {

    private static final String TAG = "My_Test";
    // 加载进度条
    private Dialog mLoadingDialog = null;

    protected T mPresenter;

    protected Context mContext;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mContext = this;
        initView();
        initPresenter();
        if (mPresenter != null) {
            getLifecycle().addObserver(mPresenter);
        }
    }

    protected abstract void initView();

    protected abstract void initPresenter();

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (mPresenter != null) {
            getLifecycle().removeObserver(mPresenter);
        }
    }

    @Override
    public Context getContext() {
        return this;
    }

    @Override
    public LifecycleOwner getLifecycleOwner() {
        return this;
    }

    @Override
    public void showMessage(String message) {
        Toast.makeText(this, message, Toast.LENGTH_SHORT).show();
    }

    @Override
    public void showMessage(int messageResId) {
        showMessage(getString(messageResId));
    }



    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
    }
    @Override
    protected void onResume() {
        super.onResume();
    }

    @Override
    public Activity getActivity() {
        return this;
    }

    public void onBack(View view) {
        finish();
    }
}

File: app/src/main/java/com/hjw/fundplan/base/BaseActivity.kt
package com.hjw.fundplan.base

import android.app.Activity
import android.app.Dialog
import android.content.Context
import android.os.Build
import android.os.Bundle
import android.view.View
import android.view.ViewGroup
import android.view.Window
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.core.content.ContextCompat
import androidx.lifecycle.LifecycleOwner
import com.hjw.fundplan.R
import com.hjw.fundplan.base.IContract.IPresenter
import com.hjw.fundplan.base.IContract.IView
import com.hjw.fundplan.util.AppUtils
import org.greenrobot.eventbus.EventBus

abstract class BaseActivity<T : IPresenter<out IView?>?> : AppCompatActivity(),
    IView {
    // 加载进度条
    private val mLoadingDialog: Dialog? = null
    protected var mPresenter: T? = null
    protected var mContext: Context? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        AppUtils.setStatusBarColor(this,R.color.light_menu_header)
        setContentView(getLayoutId())
        if (isRegisterEventBus()) {
            EventBus.getDefault().register(this)
        }
        mContext = this
        if (findViewById<Toolbar>(R.id.id_drawer_layout_toolbar) != null) {
            setSupportActionBar(findViewById<Toolbar>(R.id.id_drawer_layout_toolbar))
            if (setNavigation()) {
                findViewById<Toolbar>(R.id.id_drawer_layout_toolbar).setNavigationIcon(R.drawable.back)
                findViewById<Toolbar>(R.id.id_drawer_layout_toolbar).setNavigationOnClickListener { finish() }
            }
        }
        initPresenter()
        initView()
        if (mPresenter != null) {
            lifecycle.addObserver(mPresenter!!)
        }
    }


    protected abstract fun initView()
    protected abstract fun getLayoutId(): Int
    open fun setNavigation(): Boolean = true
    open fun isRegisterEventBus(): Boolean = false
    protected abstract fun initPresenter()
    override fun onDestroy() {
        super.onDestroy()
        if (isRegisterEventBus()) {
            EventBus.getDefault().unregister(this)
        }
        if (mPresenter != null) {
            lifecycle.removeObserver(mPresenter!!)
        }
    }

    override val context: Context
        get() = this

    override val lifecycleOwner: LifecycleOwner
        get() = this

    override fun showMessage(message: String?) {
        Toast.makeText(this, message, Toast.LENGTH_SHORT).show()
    }

    override fun showMessage(messageResId: Int) {
        showMessage(getString(messageResId))
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
    }

    override fun onResume() {
        super.onResume()
    }

    override val activity: Activity
        get() = this

    fun onBack(view: View?) {
        finish()
    }

    companion object {
        private const val TAG = "My_Test"
    }
}